src/BaseClient.php

```diff
@@ -108,53 +136,137 @@ public function getToken(): ?array
+            'grant_type' => 'refresh_token',
+            'refresh_token' => $refreshToken->encode(),
+        ]);
+
+        $tokenResponse = $this->requestToken($clientId, $clientSecret, $tokenRequest);
+
+        return new JWTToken($tokenResponse->refresh_token);
+    }
+
+    /**
+     * Executes the request to the token endpoint.
+     *
+     * @param string $clientId The client ID to use for authentication.
+     * @param string $clientSecret The client secret to use for authentication.
+     * @param TokenRequest $token The token request.
+     * @param $expectedScope ?string The expected scope of the token.
+     * @return TokenResponse The token response.
+     *
+     * @throws ConnectException when an error occurred in the HTTP connection.
+     * @throws ResponseException when an unexpected response was received.
+     * @throws UnauthorizedException when authentication failed.
+     * @throws RateLimitException when the throttling limit has been reached for the API user.
+     * @throws Exception when something unexpected went wrong.
+     */
+    protected function requestToken(string $clientId, string $clientSecret, TokenRequest $token, ?string $expectedScope = null): TokenResponse
     {
         $credentials = base64_encode(sprintf('%s:%s', $clientId, $clientSecret));
         $response = $this->rawRequest('POST', static::API_TOKEN_URI, [
             'headers' => [
                 'Accept' => 'application/json',
                 'Authorization' => sprintf('Basic %s', $credentials)
             ],
-            'query' => [
-                'grant_type' => 'client_credentials'
-            ]
+            'query' => $token->toArray()
         ]);
 
-        $token = $this->jsonDecodeBody($response);
-        $this->validateToken($token);
+        $responseTypes = [
+            '200' => TokenResponse::class
+        ];
+
+        $tokenResponse = $this->decodeResponse($response, $responseTypes, static::API_TOKEN_URI);
+        $this->validateTokenResponse($tokenResponse, $expectedScope);
+        $this->accessToken = new JWTToken($tokenResponse->access_token);
 
-        $token['expires_at'] = time() + $token['expires_in'] ?? 0;
-        $this->token = $token;
+        return $tokenResponse;
     }
 
     /**
-     * @param $token Token data from HTTP response.
+     * @param $tokenResponse TokenResponse data from HTTP response.
+     * @param $expectedScope ?string The expected scope of the token.
      *
      * @throws ResponseException when the token data is invalid.
      */
-    protected function validateToken($token): void
+    protected function validateTokenResponse(TokenResponse $tokenResponse, ?string $expectedScope = null): void
     {
-        if (! is_array($token)) {
-            throw new ResponseException('Token is not an array');
-        }
 
-        if (empty($token['access_token'])) {
+        if ($tokenResponse->access_token === null || $tokenResponse->access_token === '') {
```

Comment:
Ik check dit meestal door het volgende te doen:

```suggestion
        if (trim($tokenResponse->access_token ?? '') === '') {
```

Dat doet hetzelfde, maar is net iets makkelijker te lezen (voor mij persoonlijk).